### PR TITLE
Code fix 1678

### LINKED ISSUE #1678: Code improvement in example of "WeatherComponent2"
I suggest to change the code in example of "WeatherComponent2" as follows:

old:

```
private void RaiseEvent(GetWeatherCompletedEventArgs e)
    {
        if (GetWeatherCompleted != null)
            GetWeatherCompleted(this, e);
    }

```

new:
```
private void RaiseEvent(GetWeatherCompletedEventArgs e)
    {
        GetWeatherCompleted?.Invoke(this, e);
    }
```

Reason: Safer (between the two accesses to `GetWeatherCompleted ` someone could unsubscribe and cause a `NullReferenceException`) and more compact syntax.

---
#### Document Details

⚠ *Do not edit this section. It is required for learn.microsoft.com ➟ GitHub issue linking.*

* ID: b0bc9d86-b893-7c9f-d832-3c3026f5278e
* Version Independent ID: 5d75b64a-62d1-b0f9-b8fd-c84f60318808
* Content: [Threading Model - WPF .NET Framework](https://learn.microsoft.com/en-us/dotnet/desktop/wpf/advanced/threading-model?view=netframeworkdesktop-4.8)
* Content Source: [dotnet-desktop-guide/framework/wpf/advanced/threading-model.md](https://github.com/dotnet/docs-desktop/blob/main/dotnet-desktop-guide/framework/wpf/advanced/threading-model.md)
* Product: **dotnet-framework**
* Technology: **dotnet-wpf**
* GitHub Login: @adegeo
* Microsoft Alias: **adegeo**

---
[Associated WorkItem - 153648](https://dev.azure.com/msft-skilling/Content/_workitems/edit/153648)

### LINKED ISSUE #1691: Code example is missing information (WPF - Threading Model)
The code example in the section [Handling a Blocking Operation with a Background Thread](https://learn.microsoft.com/en-us/dotnet/desktop/wpf/advanced/threading-model?view=netframeworkdesktop-4.8&redirectedfrom=MSDN#handling-a-blocking-operation-with-a-background-thread) uses a dispatcher in the main UI thread to invoke a function, but the main UI thread isn't defined in the code (or other code examples on the page).

The thread object is called `tomorrowsWeather` and searching the page (Ctrl+F) for that keyword shows that the object isn't mentioned anywhere else (except in a excerpt from the same code example). The main UI thread is a major part of the section's topic, so it needs to be shown/explained.

The code example for convenience:
```csharp
using System;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Media;
using System.Windows.Media.Animation;
using System.Windows.Media.Imaging;
using System.Windows.Shapes;
using System.Windows.Threading;
using System.Threading;

namespace SDKSamples
{
    public partial class Window1 : Window
    {
        // Delegates to be used in placking jobs onto the Dispatcher.
        private delegate void NoArgDelegate();
        private delegate void OneArgDelegate(String arg);

        // Storyboards for the animations.
        private Storyboard showClockFaceStoryboard;
        private Storyboard hideClockFaceStoryboard;
        private Storyboard showWeatherImageStoryboard;
        private Storyboard hideWeatherImageStoryboard;

        public Window1(): base()
        {
            InitializeComponent();
        }

        private void Window_Loaded(object sender, RoutedEventArgs e)
        {
            // Load the storyboard resources.
            showClockFaceStoryboard =
                (Storyboard)this.Resources["ShowClockFaceStoryboard"];
            hideClockFaceStoryboard =
                (Storyboard)this.Resources["HideClockFaceStoryboard"];
            showWeatherImageStoryboard =
                (Storyboard)this.Resources["ShowWeatherImageStoryboard"];
            hideWeatherImageStoryboard =
                (Storyboard)this.Resources["HideWeatherImageStoryboard"];
        }

        private void ForecastButtonHandler(object sender, RoutedEventArgs e)
        {
            // Change the status image and start the rotation animation.
            fetchButton.IsEnabled = false;
            fetchButton.Content = "Contacting Server";
            weatherText.Text = "";
            hideWeatherImageStoryboard.Begin(this);

            // Start fetching the weather forecast asynchronously.
            NoArgDelegate fetcher = new NoArgDelegate(
                this.FetchWeatherFromServer);

            fetcher.BeginInvoke(null, null);
        }

        private void FetchWeatherFromServer()
        {
            // Simulate the delay from network access.
            Thread.Sleep(4000);

            // Tried and true method for weather forecasting - random numbers.
            Random rand = new Random();
            String weather;

            if (rand.Next(2) == 0)
            {
                weather = "rainy";
            }
            else
            {
                weather = "sunny";
            }

            // Schedule the update function in the UI thread.
            tomorrowsWeather.Dispatcher.BeginInvoke(
                System.Windows.Threading.DispatcherPriority.Normal,
                new OneArgDelegate(UpdateUserInterface),
                weather);
        }

        private void UpdateUserInterface(String weather)
        {
            //Set the weather image
            if (weather == "sunny")
            {
                weatherIndicatorImage.Source = (ImageSource)this.Resources[
                    "SunnyImageSource"];
            }
            else if (weather == "rainy")
            {
                weatherIndicatorImage.Source = (ImageSource)this.Resources[
                    "RainingImageSource"];
            }

            //Stop clock animation
            showClockFaceStoryboard.Stop(this);
            hideClockFaceStoryboard.Begin(this);

            //Update UI text
            fetchButton.IsEnabled = true;
            fetchButton.Content = "Fetch Forecast";
            weatherText.Text = weather;
        }

        private void HideClockFaceStoryboard_Completed(object sender,
            EventArgs args)
        {
            showWeatherImageStoryboard.Begin(this);
        }

        private void HideWeatherImageStoryboard_Completed(object sender,
            EventArgs args)
        {
            showClockFaceStoryboard.Begin(this, true);
        }
    }
}
```

---
#### Document Details

⚠ *Do not edit this section. It is required for learn.microsoft.com ➟ GitHub issue linking.*

* ID: b0bc9d86-b893-7c9f-d832-3c3026f5278e
* Version Independent ID: 5d75b64a-62d1-b0f9-b8fd-c84f60318808
* Content: [Threading Model - WPF .NET Framework](https://learn.microsoft.com/en-us/dotnet/desktop/wpf/advanced/threading-model?view=netframeworkdesktop-4.8&redirectedfrom=MSDN)
* Content Source: [dotnet-desktop-guide/framework/wpf/advanced/threading-model.md](https://github.com/dotnet/docs-desktop/blob/main/dotnet-desktop-guide/framework/wpf/advanced/threading-model.md)
* Product: **dotnet-framework**
* Technology: **dotnet-wpf**
* GitHub Login: @adegeo
* Microsoft Alias: **adegeo**

---
[Associated WorkItem - 153649](https://dev.azure.com/msft-skilling/Content/_workitems/edit/153649)

### PR DESCRIPTION
**Pull request by adegeo**
- Rewrote the intro note and the **Overview and the dispatcher** section to focus on `InvokeAsync` instead of `BeginInvoke`.
- Rewrote the code for the 3 examples.
- Minor touchups to wording, **but I didn't rewrite the article**.
- New screenshots; kept old diagrams

Fixes #1691
Fixes #1678

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [dotnet-desktop-guide/framework/wpf/advanced/threading-model.md](https://github.com/dotnet/docs-desktop/blob/91522e79edd4b411dfbea657d12866396eaf2ae5/dotnet-desktop-guide/framework/wpf/advanced/threading-model.md) | [Threading Model](https://review.learn.microsoft.com/en-us/dotnet/desktop/wpf/advanced/threading-model?branch=pr-en-us-1704&view=netframeworkdesktop-4.8) |

### CODE REVIEW

File: dotnet-desktop-guide/framework/wpf/advanced/snippets/threading-model/csharp/Weather.xaml.cs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+            // Asynchronously fetch the weather forecast on a different thread and pause this code.
+            string weather = await Task.Run(FetchWeatherFromServer);
+
+            // After async data returns, process it...
+            // Set the weather image
+            if (weather == "sunny")
+                weatherIndicatorImage.Source = (ImageSource)Resources["SunnyImageSource"];
+
+            else if (weather == "rainy")
+                weatherIndicatorImage.Source = (ImageSource)Resources["RainingImageSource"];
```
Comment (IEvangelist) [inline dotnet-desktop-guide/framework/wpf/advanced/snippets/threading-model/csharp/Weather.xaml.cs:33]:
```suggestion
            weatherIndicatorImage.Source = weather switch
            {
                "sunny" => (ImageSource)Resources["SunnyImageSource"],
                _ =>  (ImageSource)Resources["RainingImageSource"]
            };
```

File: dotnet-desktop-guide/framework/wpf/advanced/snippets/threading-model/csharp/Weather.xaml.cs [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+            
+            //Update UI text
+            fetchButton.IsEnabled = true;
+            fetchButton.Content = "Fetch Forecast";
+            weatherText.Text = weather;
+        }
+        //</button>
+
+        //<fetch_weather>
+        private async Task<string> FetchWeatherFromServer()
```
Comment (IEvangelist) [inline dotnet-desktop-guide/framework/wpf/advanced/snippets/threading-model/csharp/Weather.xaml.cs:47]:
```suggestion
        private async Task<string> FetchWeatherFromServerAsync()
```

File: dotnet-desktop-guide/framework/wpf/advanced/snippets/threading-model/csharp/Weather.xaml.cs [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+        private async void FetchButton_Click(object sender, RoutedEventArgs e)
+        {
+            // Change the status image and start the rotation animation.
+            fetchButton.IsEnabled = false;
+            fetchButton.Content = "Contacting Server";
+            weatherText.Text = "";
+            ((Storyboard)Resources["HideWeatherImageStoryboard"]).Begin(this);
+
+            // Asynchronously fetch the weather forecast on a different thread and pause this code.
+            string weather = await Task.Run(FetchWeatherFromServer);
```
Comment (IEvangelist) [inline dotnet-desktop-guide/framework/wpf/advanced/snippets/threading-model/csharp/Weather.xaml.cs:25]:
```suggestion
            string weather = await FetchWeatherFromServerAsync();
```

File: dotnet-desktop-guide/framework/wpf/advanced/snippets/threading-model/csharp/MultiWindow.xaml.cs [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    {
+        public MultiWindow() =>
+            InitializeComponent();
+
+        private void Window_Loaded(object sender, RoutedEventArgs e) =>
+            ThreadStatusItem.Content = $"Thread ID: {Thread.CurrentThread.ManagedThreadId}";
+
+        //<delay>
+        private void StartStopButton_Click(object sender, RoutedEventArgs e) =>
+            Task.Delay(TimeSpan.FromSeconds(5)).Wait();
```
Comment (IEvangelist) [inline dotnet-desktop-guide/framework/wpf/advanced/snippets/threading-model/csharp/MultiWindow.xaml.cs:19]: Hmmm, why wouldn't you do `async / await` here?

Comment (adegeo) [reply dotnet-desktop-guide/framework/wpf/advanced/snippets/threading-model/csharp/MultiWindow.xaml.cs:19]: This is demonstrating blocking the thread so you can see the windows that were launched under other threads continue to operate as normal.

File: dotnet-desktop-guide/framework/wpf/advanced/snippets/threading-model/csharp/Weather.xaml.cs (hunk 1 above)
Comment (adegeo) [reply dotnet-desktop-guide/framework/wpf/advanced/snippets/threading-model/csharp/Weather.xaml.cs:33]: Can't do this in .NET framework.

File: dotnet-desktop-guide/framework/wpf/advanced/snippets/threading-model/csharp/AssemblyInfo.cs [hunk 5]
```diff
@@ -0,0 +1,10 @@
+using System.Windows;
+
+[assembly:ThemeInfo(
+    ResourceDictionaryLocation.None,            //where theme specific resource dictionaries are located
+                                                //(used if a resource is not found in the page,
+                                                // or application resource dictionaries)
+    ResourceDictionaryLocation.SourceAssembly   //where the generic resource dictionary is located
+                                                //(used if a resource is not found in the page,
+                                                // app, or any theme specific resource dictionaries)
+)]
```
[the remark this answers is not in the document]
Comment (adegeo) [reply dotnet-desktop-guide/framework/wpf/advanced/snippets/threading-model/csharp/AssemblyInfo.cs:10]: This is just what the template generates, so modification isn't required.

**Comment by adegeo** [comment]
@Micke3rd  Ack yes... And some other variables were changed via the feedback that weren't changed in the description. I'll run through the article and fix these up.

**Pull request merged**